File: forms/medical-operation-note/back-end-with-loco/medical_operation_note/src/engine/count_rules.rs
//! Swab / needle / instrument count rules.
//!
//! Any unresolved discrepancy forces composite risk to at least High-risk
//! (or Critical when the discrepancy is suspected to be a retained item).

use super::types::{CompositeRisk, FiredRule};

/// Evaluate.
pub fn evaluate(
    swab_agreed: bool,
    needle_agreed: bool,
    instrument_agreed: bool,
    retained_foreign_body: bool,
) -> (CompositeRisk, bool, Option<FiredRule>) {
    let counts_agreed = swab_agreed && needle_agreed && instrument_agreed;
    if retained_foreign_body {
        return (
            CompositeRisk::Critical,
            counts_agreed,
            Some(FiredRule {
                rule_id: "count.retained".into(),
                instrument: "counts".into(),
                band: "critical".into(),
                category: "retained-foreign-body".into(),
                description: "Retained foreign body declared".into(),
            }),
        );
    }
    if !counts_agreed {
        return (
            CompositeRisk::HighRisk,
            counts_agreed,
            Some(FiredRule {
                rule_id: "count.discrepancy".into(),
                instrument: "counts".into(),
                band: "high-risk".into(),
                category: "count-discrepancy".into(),
                description: "Swab/needle/instrument count not agreed at sign-out".into(),
            }),
        );
    }
    (CompositeRisk::Routine, counts_agreed, None)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn all_agreed_no_retained() {
        let (r, agreed, rule) = evaluate(true, true, true, false);
        assert_eq!(r, CompositeRisk::Routine);
        assert!(agreed);
        assert!(rule.is_none());
    }

    #[test]
    fn one_disagreed_high_risk() {
        let (r, agreed, rule) = evaluate(true, false, true, false);
        assert_eq!(r, CompositeRisk::HighRisk);
        assert!(!agreed);
        assert_eq!(rule.unwrap().rule_id, "count.discrepancy");
    }

    #[test]
    fn retained_critical() {
        let (r, _, rule) = evaluate(true, true, true, true);
        assert_eq!(r, CompositeRisk::Critical);
        assert_eq!(rule.unwrap().rule_id, "count.retained");
    }
}

File: forms/medical-operation-note/front-end-with-svelte/src/lib/engine/count-rules.ts
import type { CompositeRisk, FiredRule, OperationNote } from './types.js';

/**
 * Resolve the swab/needle/instrument-count discrepancy state at sign-out.
 *
 * - All three explicitly agreed → counts agreed; no rule fires.
 * - Any one explicitly disagreed → discrepancy: composite risk ≥ High-risk.
 * - Retained foreign body declared independently → handled in
 *   `never-event-rules.ts` (forces Critical).
 */
export function applyCountRules(data: OperationNote): {
  countsAgreed: boolean;
  discrepancyRisk: CompositeRisk;
  firedRules: FiredRule[];
} {
  const s = data.safetyCountsEbl;
  const firedRules: FiredRule[] = [];

  const anyDisagreed =
    s.swabCountAgreed === 'no' ||
    s.needleCountAgreed === 'no' ||
    s.instrumentCountAgreed === 'no';

  const allAgreed =
    s.swabCountAgreed === 'yes' &&
    s.needleCountAgreed === 'yes' &&
    s.instrumentCountAgreed === 'yes';

  if (anyDisagreed) {
    firedRules.push({
      ruleId: 'R-COUNTS-DISCREPANCY',
      instrument: 'counts',
      grade: 'discrepancy',
      category: 'safety',
      description: 'Swab / needle / instrument count discrepancy at sign-out',
    });
    return { countsAgreed: false, discrepancyRisk: 'high-risk', firedRules };
  }

  return {
    countsAgreed: allAgreed,
    discrepancyRisk: 'routine',
    firedRules,
  };
}
